eth_account/hdaccount/mnemonic.py

```diff
@@ -0,0 +1,202 @@
+        # other) then return simplified. This doesn't hold for other languages.
+        if len(matching_languages) == 2 and all("chinese" in lang for lang in matching_languages):
+            return "chinese_simplified"
+
+        # Because certain wordlists share some similar words, if we detect multiple languages
+        # that the provided mnemonic word(s) could be valid in, we have to throw
+        if len(matching_languages) > 1:
+            raise ValidationError(f"Word(s) are valid in multiple languages: {raw_mnemonic}")
+
+        (language,) = matching_languages
+        return language
+
+    def generate(self, num_words=12):
+        if num_words not in VALID_WORD_COUNTS:
+            raise ValidationError(
+                f"Invalid choice for number of words: {num_words}, should be one of "
+                f"{VALID_WORD_COUNTS}"
+            )
+        return self.to_mnemonic(os.urandom(4 * num_words // 3))  # 4/3 bytes per word
+
+    def to_mnemonic(self, entropy):
+        entropy_size = len(entropy)
+        if entropy_size not in VALID_ENTROPY_SIZES:
+            raise ValidationError(
+                f"Invalid data length {len(entropy)}, should be one of "
+                f"{VALID_ENTROPY_SIZES}"
+            )
+
+        bits = bitarray()
+        bits.frombytes(entropy)
+
+        checksum = bitarray()
+        checksum.frombytes(sha256(entropy))
+
+        # Add enough bits from the checksum to make it modulo 11 (2**11 = 2048)
+        bits.extend(checksum[:entropy_size // 4])
+        indices = tuple(ba2int(bits[i * 11: (i + 1) * 11]) for i in range(len(bits) // 11))
+        words = tuple(self.wordlist[idx] for idx in indices)
+
+        if self.language == "japanese":  # Japanese must be joined by ideographic space.
+            phrase = u"\u3000".join(words)
+        else:
+            phrase = " ".join(words)
+        return phrase
+
+    def is_mnemonic_valid(self, mnemonic):
+        words = normalize_string(mnemonic).split(" ")
+        num_words = len(words)
+
+        if num_words not in VALID_WORD_COUNTS:
+            return False
+
+        try:
+            indices = tuple(self.wordlist.index(w) for w in words)
+        except ValueError:
+            return False
+
+        encoded_seed = bitarray()
+        for idx in indices:
+            # Build bitarray from tightly packing indices (which are 11-bits integers)
+            encoded_seed.extend(int2ba(idx, length=11))
+
+        entropy_size = 4 * num_words // 3
+
+        # Checksum the raw entropy bits
+        checksum = bitarray()
+        checksum.frombytes(sha256(encoded_seed[:entropy_size * 8].tobytes()))
+
+        # Extract the stored checksum bits
+        stored_checksum = encoded_seed[entropy_size * 8:]
+
+        # Check that the stored matches the relevant slice of the actual checksum
+        return stored_checksum == checksum[:len(encoded_seed) - entropy_size * 8]
```

Comment:
I believe still should use a constant time comparison operation.  I'm not sure if there is a timing attack hiding in here but if there's any uncertainty, the constant time `secrets.compare_digest` is the tool to mitigate against that.

Reply:
Solved in 81b7748